akka-http-core/src/main/scala/akka/http/impl/engine/http2/client/PersistentConnection.scala

```diff
@@ -0,0 +1,127 @@
+  private val associationTagKey = AttributeKey[AssociationTag]("PersistentConnection.associationTagKey")
+  private val errorResponse =
+    HttpResponse(
+      StatusCodes.BadGateway,
+      entity = "The server closed the connection before delivering a response.")
+
+  private class Stage(connectionFlow: Flow[HttpRequest, HttpResponse, Any]) extends GraphStage[FlowShape[HttpRequest, HttpResponse]] {
+    val requestIn = Inlet[HttpRequest]("PersistentConnection.requestIn")
+    val responseOut = Outlet[HttpResponse]("PersistentConnection.responseOut")
+
+    val shape: FlowShape[HttpRequest, HttpResponse] = FlowShape(requestIn, responseOut)
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with StageLogging {
+      become(Unconnected)
+
+      def become(state: State): Unit = setHandlers(requestIn, responseOut, state)
+
+      trait State extends InHandler with OutHandler
+      object Unconnected extends State {
+        override def onPush(): Unit = connect()
+        override def onPull(): Unit =
+          if (!hasBeenPulled(requestIn)) // requestIn might already have been pulled when we failed and went back to Unconnected
+            pull(requestIn)
+
+      }
+
+      def connect(): Unit = {
+        val requestOut = new SubSourceOutlet[HttpRequest]("PersistentConnection.requestOut")
+        val responseIn = new SubSinkInlet[HttpResponse]("PersistentConnection.responseIn")
+
+        Source.fromGraph(requestOut.source)
+          .via(connectionFlow)
+          .runWith(responseIn.sink)(subFusingMaterializer)
+
+        become(new Connected(requestOut, responseIn))
+      }
+
+      class Connected(
+        requestOut: SubSourceOutlet[HttpRequest],
+        responseIn: SubSinkInlet[HttpResponse]
+      ) extends State {
+        private var ongoingRequests: Map[AssociationTag, Map[AttributeKey[_], RequestResponseAssociation]] = Map.empty
+        responseIn.pull()
+
+        requestOut.setHandler(new OutHandler {
+          override def onPull(): Unit =
+            if (!isAvailable(requestIn)) pull(requestIn)
+            else dispatchRequest(grab(requestIn))
+
+          override def onDownstreamFinish(): Unit = onDisconnected()
+        })
+        responseIn.setHandler(new InHandler {
+          override def onPush(): Unit = {
+            val response = responseIn.grab()
+            val tag = response.attribute(associationTagKey).get
+            require(ongoingRequests.contains(tag))
+            ongoingRequests -= tag
+            push(responseOut, response)
+          }
+
+          override def onUpstreamFinish(): Unit = onDisconnected()
+          override def onUpstreamFailure(ex: Throwable): Unit = onDisconnected() // FIXME: log error
+        })
+        def onDisconnected(): Unit = {
+          emitMultiple[HttpResponse](responseOut, ongoingRequests.values.map(errorResponse.withAttributes(_)).toVector, () => setHandler(responseOut, Unconnected))
```

Comment:
`onDisconnected` is currently only used for completion signals from the client implementation. So, for the server-side, the only case we might handle better is when the server sends a GOAWAY which our client implementation would interpret as "accept no new requests but might still send responses" and therefore only cancels the request stream but does not yet complete the response stream.

Is that what you mean?

In any case, we have not implemented that behavior for GOAWAY in the HTTP/2 stack so far.

Or is it about the user completing the request stream? In that case, we might just document for now that you must not complete the request stream if you still want to read responses (but it would be nice indeed to support the use case you mentioned).

Reply:
+1 to explicitly mention on the docs that completing the stream doesn't wait for in-flight requests to complete. That is the behavior in non-persistent connections already but it won't harm if we specify it's the same on persistent connections.

Reply:
Just saw there's #3594 to implement the improvement.